internal/services/report_processor.go

```diff
@@ -75,122 +75,138 @@ func ProcessReport(msg *kafka.Message) {
-			usage_data_byte, err := kruize.Update_results(experiment_name, k8s_object)
-			if err != nil {
-				log.Error(err)
-				continue
-			}
+				experiment_name := utils.GenerateExperimentName(
+					kafkaMsg.Metadata.Org_id,
+					kafkaMsg.Metadata.Source_id,
+					kafkaMsg.Metadata.Cluster_uuid,
+					namespace,
+					k8s_object_type,
+					k8s_object_name,
+				)
+
+				if workload_metrics, err := model.GetWorkloadMetricsForTimestamp(experiment_name, interval_end); err != nil {
+					log.Errorf("Error while checking for workload_metrics record: %s", err)
+					continue
+				} else if !reflect.ValueOf(workload_metrics).IsZero() {
+					log.Debugf("worload_metrics table already has data for interval_end time: %v.", interval_end)
+					continue
+				}
 
-			for _, container := range usage_data_byte[0].Kubernetes_objects[0].Containers {
-				container_usage_metrics, err := json.Marshal(container.Metrics)
+				container_names, err := kruize.Create_kruize_experiments(experiment_name, k8s_object)
 				if err != nil {
-					log.Errorf("Unable to marshal container usage data: %v", err)
+					log.Error(err)
+					continue
 				}
-				workload_metric := model.WorkloadMetrics{
-					WorkloadID:    workload.ID,
-					ContainerName: container.Container_name,
-					IntervalStart: interval_start,
-					IntervalEnd:   interval_end,
-					UsageMetrics:  container_usage_metrics,
+
+				// Create workload entry into the table.
+				workload := model.Workload{
+					ClusterID:       cluster.ID,
+					ExperimentName:  experiment_name,
+					Namespace:       namespace,
+					WorkloadType:    workload.WorkloadType(k8s_object_type),
+					WorkloadName:    k8s_object_name,
+					Containers:      container_names,
+					MetricsUploadAt: time.Now(),
 				}
-				if err := workload_metric.CreateWorkloadMetrics(); err != nil {
-					log.Errorf("unable to add record to workload_metrics table: %v. Error: %v", workload_metric, err)
+				if err := workload.CreateWorkload(); err != nil {
+					log.Errorf("unable to get or add record to workloads table: %v. Error: %v", workload, err)
 					continue
 				}
-			}
 
-			waittime, err := strconv.Atoi(cfg.KruizeWaitTime)
-			if err != nil {
-				log.Error(err)
-			}
-			// Sending list_of_experiments to rosocp.kruize.experiments topic.
-			experimentEventMsg := types.ExperimentEvent{
-				WorkloadID:          workload.ID,
-				Experiment_name:     experiment_name,
-				K8s_object_name:     k8s_object[0]["k8s_object_name"].(string),
-				K8s_object_type:     k8s_object[0]["k8s_object_type"].(string),
-				Namespace:           k8s_object[0]["namespace"].(string),
-				Fetch_time:          time.Now().UTC().Add(time.Second * time.Duration(waittime)),
-				Monitoring_end_time: interval_end.String(),
-				K8s_object:          k8s_object,
-				Attempt:             1,
-				Kafka_request_msg:   kafkaMsg,
-			}
+				usage_data_byte, err := kruize.Update_results(experiment_name, k8s_object)
+				if err != nil {
+					log.Error(err)
+					continue
```

Comment:
@patilsuraj767 Should we attempt this continue post some secs of sleep, considering API requests fail?

Reply:
We cannot add sleep in `if block` because API call has been finished until code reaches to `if block`. In our code base everything work synchronously(http call is not async) so I don't see benefit in adding sleep.

Reply:
IIUC, the flow reaches `continue` only if there is an error, shall we add some code that checks if the API url responds and then does a `continue`?

Reply:
Actual http request does not happens from here.  `kruize.Update_results` is a custom function defined - [here](https://github.com/RedHatInsights/ros-ocp-backend/blob/main/internal/utils/kruize/kruize_api.go#L91) If have added a response check on [line 114](https://github.com/RedHatInsights/ros-ocp-backend/blob/main/internal/utils/kruize/kruize_api.go#L114).